decoder.py

```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Transformer Model
+
+# Import libraries
+
+# In[47]:
+
+
+import torch
+import torch.nn as nn
+
+
+# The embedding layer
+
+# In[65]:
+
+
+class EmbeddingLayer(nn.Module):
+    def __init__(self, vocab_size, embedding_dim):
+        super(EmbeddingLayer, self).__init__()
+        self.embed = nn.Embedding(vocab_size, embedding_dim)
+        self.embedding_dim = embedding_dim
+        
+    def positional_embedding(self, text):
+        batch_size = text.size(0)
+        sequence_length = text.size(1) 
+        embedding_dim = self.embedding_dim
+        position = torch.arange(0, sequence_length).unsqueeze(1).float()
+        div_term = torch.exp(torch.arange(0, embedding_dim, 2).float() * -(torch.log(torch.tensor(10000.0)) / embedding_dim))
+        encoding = torch.zeros(sequence_length, embedding_dim)
+        encoding[:, 0::2] = torch.sin(position * div_term)
+        encoding[:, 1::2] = torch.cos(position * div_term[:embedding_dim // 2])
+        return encoding
+
+    def forward(self, text):
+        embedded_text = self.embed(text)
+        positionally_embedded_text = embedded_text + self.positional_embedding(text)
+        return positionally_embedded_text
+
+
+# In[79]:
+
+
+# emb = EmbeddingLayer(8, 5)
+# print(emb(torch.tensor([[1, 3, 5, 7]])))
+
+
+# (Masked) Multi-head attention
+
+# In[ ]:
+
+
+class MultiHeadAttention(nn.Module):
+    def __init__(self, d_model, num_heads, forward_masked):
+        super(MultiHeadAttention, self).__init__()
+        self.d_model = d_model
```

Comment:
how about we are consistent across the codebase with naming convention for dimenality?

d_model? head_dim?

Reply:
ff_dim?

Reply:
embedding_dim?

Reply:
Fair enough, maybe later.